docs/cloud-integration.md

```diff
@@ -0,0 +1,512 @@
+using a URL to the data in methods like `SparkContext.textFile()` to read data, 
+`saveAsTextFile()` to write it back.
+
+
+Because object stores are viewed by Spark as filesystems, object stores can
+be used as the source or destination of any spark work —be it batch, SQL, DataFrame,
+Streaming or something else.
+
+The steps to do so are as follows
+
+1. Use the full URI to refer to a bucket, including the prefix for the client-side library
+to use. Example: `s3a://landsat-pds/scene_list.gz`
+1. Have the Spark context configured with the authentication details of the object store.
+In a YARN cluster, this may also be done in the `core-site.xml` file.
+
+
+## <a name="output"></a>Object Stores as a substitute for HDFS
+
+As the examples show, you can write data to object stores. However, that does not mean
+That they can be used as replacements for a cluster-wide filesystem.
+
+The full details are covered in [Cloud Object Stores are Not Real Filesystems](#cloud_stores_are_not_filesystems).
+
+The brief summary is:
+
+| Object Store Connector      |  Replace HDFS? |
+|-----------------------------|--------------------|
+| `s3a://` `s3n://`  from the ASF   | No  |
+| Amazon EMR `s3://`          | Yes |
+| Microsoft Azure `wasb://`   | Yes |
+| OpenStack `swift://`        | No  |
+
+It is possible to use any of the object stores as a destination of work, i.e. use
+`saveAsTextFile()` or `save()` to save data there, but the commit process may be slow
+and, unreliable in the presence of failures.
+
+It is faster and safer to use the cluster filesystem as the destination of Spark jobs,
+using that data as the data for follow-on work. The final results can
+be persisted in to the object store using `distcp`.
+
+#### <a name="checkpointing"></a>Spark Streaming and object stores
+
+Spark Streaming can monitor files added to object stores, by
+creating a `FileInputDStream` DStream monitoring a path under a bucket through
+`StreamingContext.textFileStream()`.
+
+
+1. The time to scan for new files is proportional to the number of files
+under the path —not the number of *new* files, and that it can become a slow operation.
+The size of the window needs to be set to handle this.
+
+1. Files only appear in an object store once they are completely written; there
+is no need for a worklow of write-then-rename to ensure that files aren't picked up
+while they are still being written. Applications can write straight to the monitored directory.
+
+1. Streams should only be checkpointed to an object store considered compatible with
+HDFS. Otherwise the checkpointing will be slow and potentially unreliable.
+
+### Recommended settings for writing to object stores
+
+Here are the settings to use when writing to object stores. This uses the "version 2" algorithm
+for committing files —which does less renaming than the v1 algorithm. Speculative execution is
+disabled to avoid multiple writers corrupting the output.
```

Comment:
MRv1 is just way slower, with more renaming.
Speculation is potentially dangerous due to the way rename() is used for some atomic bits of the commit. It's why the Direct Committer always turned off speculation, to avoid > 1 executor trying to write to the same destination directory. 

That said, if it is up to the job committer to do the final commit action, then speculation isn't so risky. After all: the commit algorithm needs to be able to handle executor failure anyway, and speculation is similar, but actually easier, as theres a commit protocol to help abort work.